### DISCUSSION TITLE
Cloudflare - pmtiles file not loading

### DISCUSSION THREAD

**Discussion by kbatchu** [protomaps/PMTiles#284] [Q&A] [answered]
I uploaded a pmtiles file to an r2 instance on Cloudflare, configured a custom domain and created a worker with the code provided in the Protomaps docs for Cloudflare integration. But I am getting this error while loading the file through a web app.
![image](https://github.com/protomaps/PMTiles/assets/34347771/e5247b62-57ca-4bb0-ab11-160b4d4c510b)
I checked the [url](https://pmtiles.kbatchu.dev/acs5yr2021popdotdensity500highresv2/0/0/0.mvt) in postman and it returns a small chunk (148KB) which looks right.
![image](https://github.com/protomaps/PMTiles/assets/34347771/a42d2448-2ced-4c7f-871f-8a918d9c9da2)
The returned content looks appropriate.
![image](https://github.com/protomaps/PMTiles/assets/34347771/749b0457-b87d-40b9-931c-c55ee624582b)

I checked this url in the PMTiles viewer and get the same error message. What seems to be the problem?
Thanks,
Kiran

**Comment by bdon** [comment, marked as answer]
If you are using the cloudflare worker integration, than your endpoint /z/x/y.mvt is now transformed into a ZXY endpoint, not a .pmtiles anymore, so you should specify that ZXY URL in your frontend code; you don't need any PMTiles related libraries on the frontend at all.

**Reply by kbatchu** [reply]
That was the clarification I needed. Works great now. Thank you so much, Brandon! 
Will the Cloudflare function also work for raster pmtiles? Or will it work for only vector pmtiles?

**Reply by bdon** [reply]
You can store raster images in pmtiles as well. You will need to address them by `.png` or `.jpg`, whatever tile type is specified in in the archive.

However, it does not do any vector to raster rasterization (you need a renderer like MapLibre native for that)

**Reply by kbatchu** [reply]
Got it. I am using OpenLayers though...for now.

**Reply by kbatchu** [reply]
So for raster pmtiles, I am trying to use this example in OpenLayers: https://openlayers.org/en/latest/examples/pmtiles-elevation.html. In this example they use the 'pmtiles' JS library and use it to get the tiles. 
const tiles = new pmtiles.PMTiles(
  'https://pub-9288c68512ed46eca46ddcade307709b.r2.dev/protomaps-sample-datasets/terrarium_z9.pmtiles'
);

But since I am using the Cloudflare worker code, I am not clear where my cloudflare url "https://pmtiles.kbatchu.dev/Georgia_CensusTracts_PopDotDensity_Raster/{z}/{x}/{y}.png" needs to go since OpenLayer's DataTileSource does not take a URL but a 'loader' function.
Any tips or suggestions?
Thanks,

**Reply by bdon** [reply]
For {z}/{x}/{y} URLs as exposed by the worker you can use any of the ZXY openlayers examples https://openlayers.org/en/latest/examples/xyz.html

**Reply by kbatchu** [reply]
Thank you so much. The raster PMTiles file loads now. One thing I noticed is that I get these errors in the browser console when loading the raster file but not for the vector file. The yellow polygons are vector and the gray dots are raster. Is there any setting in Cloudflare that I need to tweak for raster files?
![image](https://github.com/protomaps/PMTiles/assets/34347771/8a706afd-cd0a-4de7-bebd-c7a9d32cbb73)

**Reply by bdon** [reply]
Is your raster file only a limited extent, the square in the middle? It could be that your client is asking for tiles that do not exist in the PMTiles raster archive and getting a 404; one way to solve this is to pass a bounds to your OpenLayers ZXY tile source so it does not request tiles outside. 

It could be the case we should make Cloudflare return `204 no content` instead

**Reply by kbatchu** [reply]
Yes, the raster file is just for the state of Georgia (gray dots), while the vector file is the census tracts (shades of yellow) for the entire USA. Here is a zoomed out view. 
![image](https://github.com/protomaps/PMTiles/assets/34347771/578b2a01-d658-437c-b10a-389058ef79f1)
Here is the OpenLayers code that loads the raster and vector files. What needs to be tweaked in the Cloudflare worker to avoid these errors?. 
![code](https://github.com/protomaps/PMTiles/assets/34347771/750ac1e3-6fe0-4c74-ac9b-4540abc5374a)
![code1](https://github.com/protomaps/PMTiles/assets/34347771/1228fb61-244c-4562-a30b-b082c894511e)

**Reply by kbatchu** [reply]
When I load the same raster pmtiles file stored on my local drive using localhost, using 'ol-pmtiles' library, I do not see these error messages in the browser console. So I am assuming these errors need to be handled in the cloudflare worker code?

**Reply by bdon** [reply]
When creating [VectorTileSource](https://openlayers.org/en/latest/apidoc/module-ol_source_VectorTile.html) set `extent` based on the bounds fetched from here: https://pmtiles.kbatchu.dev/Georgia_CensusTracts_PopDotDensity_100_MBTiles_Raster.json

that will be more efficient than getting back lots of empty tiles

**Reply by kbatchu** [reply]
Just wanted to make sure if I understood correctly. So by setting the extent of the vector (census tracts) layer to the extent of the Raster layer, we are essentially 'clipping' the extent of the vector layer to the raster layer. Is this correct?

**Reply by bdon** [reply]
not exactly - the extents usually do not affect the visual display of an individual tile, only what tiles are loaded. example: the root tile 0/0/0 should match every possible extent, but if the backend returns geometry for the entire earth the renderer will display that unclipped

**Reply by kbatchu** [reply]
Then shouldn't I set the extent on the source of the Raster layer which is of type 'XYZ' ?

**Reply by bdon** [reply]
Yes, you are right - I mixed them up. Set the extent on the raster source for the gray dots.

**Reply by grovesNL** [reply]
@bdon 

> you don't need any PMTiles related libraries on the frontend at all

For what it's worth, this felt a bit subtle and I made the same mistake following the docs. I followed the PMTiles MapLibre setup (which adds the `protocol` for PMTiles) then the AWS guide (defining a lambda that serves XYZ tiles), but spent a while wondering why the lambda was trying to look up a XYZ tile for the initial PMTiles range request.

Maybe it would be useful to include a note about this in the AWS and Cloudflare setup? e.g., something as simple like "this will return XYZ tiles directly, so the PMTiles protocol shouldn't be used on the client using this setup" might be helpful

**Reply by bdon** [reply]
This is already outlined here: https://docs.protomaps.com/deploy/
And in the AWS setup under "make a test request": https://docs.protomaps.com/deploy/

I am hesitant to add documentation on how to use Z/X/Y URLs because that is redundant with existing MapLibre, Leaflet, etc documentation.

**Reply by grovesNL** [reply]
@bdon That's totally fair, most people will probably pick up on the URL format earlier and notice the difference (PMTiles protocol vs. XYZ depending on the guide). I wasn't familiar with PMTiles before this so I thought the same format would be used in both.

**Reply by kauly** [reply]
that's trick indeed. Can I use the `'@protomaps/basemaps'` when my source is a TileJSON or I dont need a Maplibre style in this case?

**Reply by kauly** [reply]
nevermind I'm actually having this problem and it's not related to pmtiles

https://github.com/visgl/react-map-gl/issues/2569

**Reply by sloreti** [reply]
Just voicing that I ran into the same confusion @grovesNL did. A brief mention of what one's options are for the various on web map libraries on https://docs.protomaps.com/deploy/ could be valuable. 

Thank you for all your work @bdon!

**Reply by bdon** [reply]
Up to now we've assumed the docs reader is familiar with web mapping, which uses the `{z}/{x}/{y}.ext` convention across all libraries. There is no formal specification or name for this.

I think the docs should be enhanced for an audience that isn't already experienced with web mapping. Maybe the part about Z/X/Y URLs should be larger, in a highlighted section, etc?
